### ISSUE TITLE
Class name formatting

### ISSUE THREAD

**Issue by bradlc** [bradlc/vscode-tailwindcss#55]
**Example:**

```html
<div class="xl:text-center md:text-left lg:text-right mx-auto"></div>
```

Would become:

```html
<div class="mx-auto md:text-left lg:text-right xl:text-center"></div>
```

Class names could be ordered by variant and/or CSS property.

1. What settings should be exposed for this?
2. What should be done with unknown (non-Tailwind) class names?

**Comment by octoxan** [comment]
I'd love if you could specify the ordering of properties, as everyone is very specific in their workflow once you start ordering properties a certain way, and maybe a dropdown option for non-tailwind classes to be at the start or end?

**Comment by huphtur** [comment]
2. Suggest putting those first in the order as it could contain extracted components classes.
eg: `<button class="btn bg-green-500 hover:bg-green-400 text-white">`

**Comment by kevinwhoffman** [comment]
Thanks for looking into this. I've handled CSS formatting in the past so I figured I'd offer some of that experience to this issue.

## Existing Approaches

For developers that care about the order of CSS properties, there are two predominant approaches:

1. Group by type
2. Alphabetical

Group by type makes the most sense for Tailwind because alphabetical would result in a messy mix of variant classes.

## Recommended Order

1. Non-Tailwind Classes
2. Non-variant classes grouped by type
3. Pseudo-Class Variants grouped by type
4. Responsive Variants grouped by type, escalating from smallest to largest breakpoint

## Recommended Grouping

Grouping CSS properties by type has been handled in a number of formatters such as [CSScomb](https://github.com/csscomb/csscomb.js/tree/master). You might consider borrowing the groupings from [CSScomb's default config](https://github.com/csscomb/csscomb.js/blob/f5c259fa7153f2e63789dec593f4915770631162/config/csscomb.json#L28-L371).

If you define the order in a JSON config, it may even be possible for users to override it with a custom JSON config in VS code settings.

**Comment by octoxan** [comment]
I wonder if we could borrow from how [stylelint-order](https://github.com/hudochenkov/stylelint-order) lets you customize it via json. https://gist.github.com/octoxan/1c0c24a75417f06096356e07d56be3c6

**Comment by huphtur** [comment]
[css-declaration-sorter](https://github.com/Siilwyn/css-declaration-sorter) may be of good use as well.

**Comment by MarvinMiles** [comment]
Waiting for this feature so much!

**Comment by evdama** [comment]
Has anyone done some work towards that so far? I'd love to have that thingy automated away from my thoughflow :)

**Comment by octoxan** [comment]
I'd switch to VS Code from PHP Storm if this becomes a thing lol

**Comment by christopherjohnallan** [comment]
This would be the biggest improvement I can think of for the tailwindcss experience

**Comment by huphtur** [comment]
This VSC extension takes care of all the things: https://marketplace.visualstudio.com/items?itemName=heybourn.headwind

**Comment by somebody32** [comment]
@huphtur headwind is cool but still has a lot of unsolved edge-cases: like no support for prefixed classes, missing classes from the latest TW versions, templated strings, etc. Would be really happy to see that supported by intellisense

**Comment by portikM** [comment]
Any updates on this front so far? Can't wait to see this live!

**Comment by hansfpc** [comment]
any updates?

**Comment by rynz** [comment]
What class name ordering does tailwindui use? Perhaps that could be the opinionated method of declaring class name order.

**Comment by jakback** [comment]
@bradlc Headwind seems to fix this issue. It is popular and actively maintained.

No need to reinvent the wheel, right?

**Comment by 1isten** [comment]
> @bradlc Headwind seems to fix this issue. It is popular and actively maintained.
> 
> No need to reinvent the wheel, right?

Just tried Headwind. It currently can't sort variant classes with `md:`, `lg:`, etc.. They will be treated as "custom classes" and be appended to the end. Something like `text-center sm:ml-auto md:text-left  ml-2` will be formatted into something like `ml-2 text-center sm:ml-auto md:text-left` instead of `ml-2 sm:ml-auto text-center md:text-left`, which is not good yet.

**Comment by cajakre** [comment]
@1isten The [PR](https://github.com/ryanhhhh/headwind/pull/53]) that tackled this issue was closed. I think it is personal preference, where the variant classes should be sorted.

**Comment by 1isten** [comment]
> it is personal preference

@cajakre Yep. After all there is no **correct** ordering. @bradlc 's example here just makes more sense to me. XD

**Comment by francoismassart** [comment]
I did make an eslint plugin for this classnames ordering, it reads your tailwind config and you can configure the order if you want: 
https://www.npmjs.com/package/eslint-plugin-tailwindcss

**Comment by cajakre** [comment]
@francoismassart Thanks, I will try that!

**Comment by bradlc** [comment]
Closing this as we have no plans to add class formatting to Tailwind CSS IntelliSense. We are working on an alternative class sorting solution, so look out for that in the near future 👀

**Comment by rynz** [comment]
@bradlc please reference this issue when your alternative class sorting solution is made public.

**Comment by vdanchenkov** [comment]
Alternative solution is a plugin for prettier I think https://github.com/tailwindlabs/prettier-plugin-tailwindcss.  Looks very good, but unfortunately prettier is not available for my programming language, ReScript. 

So if there is no plans to integrate it into the extension, we have to use separate extension for sorting, I think.

**Comment by joyoy96** [comment]
so there is an emacs lsp-mode https://github.com/merrickluo/lsp-tailwindcss based on https://github.com/tailwindlabs/tailwindcss-intellisense and it use https://github.com/avencera/rustywind that is inspired by https://github.com/heybourn/headwind an non prettier class sorting that is obsolete

I think this should be implemented on this extensions since its way more simpler than the prettier version
